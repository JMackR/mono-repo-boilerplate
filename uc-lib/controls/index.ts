export * from "./image"
export * from "./button"
export * from "./text"
export * from "./activity-indicator"
export * from "./separator"
export * from "./list"
export * from "./section-list"
export * from "./badge"
export * from "./toggle"
export * from "./login-button"
export * from "./layout"
export * from "./background"
export * from "./grid"
export * from "./touchable"
export * from "./view-block"
export * from "./text-entry"
export * from "./scroll-view"
export * from "./circle"
export * from "./slider"
export * from "./clickable"
export * from "./map"
export * from "./visibility-check"
export * from "./strike-through-text"
// export * from "./stripe-text-entry"
export * from "./input"
