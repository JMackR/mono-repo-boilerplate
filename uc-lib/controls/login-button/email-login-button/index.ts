export * from "./email-login-button"
