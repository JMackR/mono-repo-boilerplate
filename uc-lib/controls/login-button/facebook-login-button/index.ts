export * from "./facebook-login-button"
