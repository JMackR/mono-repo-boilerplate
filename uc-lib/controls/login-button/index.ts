export * from "./facebook-login-button"
export * from "./google-login-button"
export * from "./email-login-button"
export * from "./login-button.props"
