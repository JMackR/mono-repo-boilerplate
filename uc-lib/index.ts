export * from "./animations"
export * from "./assets"
export * from "./controls"
export * from "./themes"
export * from "./widgets"
export * from "./utilities"
export * from "./hooks"
export * from "./validators"
