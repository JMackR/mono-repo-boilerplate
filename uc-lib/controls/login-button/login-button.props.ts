import { LocalSVGSource } from "../../controls/image"
import { TextColors } from "../../themes"

export interface LoginButtonProps {
  buttonText: string
  onClick(): void
  testID?: string
  loading?: boolean
  isFullWidth?: boolean
}

// These are only used in the generic button from the specific ones, so they do not need to be exposed outside uc-lib
export interface ExtendedLoginButtonProps extends LoginButtonProps {
  style: LoginButtonStyleProps
  icon: LocalSVGSource
  showBorder?: boolean
}

export interface LoginButtonStyleProps {
  isFullWidth?: boolean
  mainBackgroundColor?: string
  borderColor?: string
  showBorderInDarkMode?: boolean
  iconBackgroundColor?: string
  iconTint?: keyof TextColors
  iconSize?: number
  pressedColor?: string
  hoverColor?: string
  textColor?: string
}
