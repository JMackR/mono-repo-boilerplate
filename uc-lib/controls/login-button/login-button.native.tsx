import React, { useRef } from "react"
import {
  StyleSheet,
  TouchableWithoutFeedback,
  TouchableNativeFeedback,
  TouchableOpacity,
  Platform,
  Text,
  View,
  Animated,
} from "react-native"
import { SVG } from "uc-lib/controls/image"
import { Margin, Overlay } from "uc-lib/controls/layout"
import { useColorTheme, useFontForTextType, useColor, TextColors } from "uc-lib/themes"
import { ExtendedLoginButtonProps } from "./login-button.props"
import { ActivityIndicator } from "../activity-indicator"
import { BackgroundPropType } from "test/react-native/__mocks__/react-native"

const ANIMATION_DURATION = 100
const TEXT_MARGIN = 2.5
const ICON_MARGIN = 0.25
const ICON_DIMENSION = 38

/*
  The default behavior is no border, a 38x38px icon with no background, light background, dark text, and granite press text.
*/
export const LoginButton = (props: ExtendedLoginButtonProps) => {
  const { onClick, testID, style, loading, isFullWidth, buttonText, icon } = props
  const pressStateAnimation = useRef(new Animated.Value(0)).current

  const colors = useColor().colors
  const mainBackgroundColor = style.mainBackgroundColor || colors.alwaysLight
  const pressedColor = style.pressedColor || colors.granite
  const { shade } = useColorTheme()
  const showBorder = style.showBorderInDarkMode === undefined || shade === "light" || style.showBorderInDarkMode
  const STYLES = StyleSheet.create({
    fullWidth: {
      width: "100%",
    },
    container: {
      justifyContent: "center",
      alignItems: "center",
      borderRadius: 4,
    },
    mainContainer: {
      flexDirection: "row",
      backgroundColor: mainBackgroundColor,
      borderColor: style.borderColor,
      borderWidth: style.borderColor === undefined ? 0 : 1,
    },
    iconContainer: {
      width: style.iconSize || ICON_DIMENSION,
      height: style.iconSize || ICON_DIMENSION,
      backgroundColor: style.iconBackgroundColor,
    },
    hideBorder: {
      borderWidth: 0,
    },
    containerPressed: {
      backgroundColor: pressedColor,
    },
    text: {
      color: style.textColor || colors.alwaysDark,
    },
  })

  const textStyles: object[] = [STYLES.text, useFontForTextType("primaryBody1")]
  const containerStyles: object[] = [STYLES.mainContainer, STYLES.container]
  const iconStyles: object[] = [STYLES.iconContainer, STYLES.container]
  if (isFullWidth) {
    containerStyles.push(STYLES.fullWidth)
  }
  if (!showBorder) {
    containerStyles.push(STYLES.hideBorder)
  }

  const getColorInterpolation = (from: string, to: string) => {
    return pressStateAnimation.interpolate({
      inputRange: [0, 1],
      outputRange: [from, to],
    })
  }
  const startPressStateColorAnimation = (targetEndValue: number) => {
    Animated.timing(pressStateAnimation, {
      toValue: targetEndValue,
      duration: ANIMATION_DURATION,
    }).start()
  }

  const Touchable = Platform.OS === "android" ? TouchableNativeFeedback : TouchableWithoutFeedback
  let background: BackgroundPropType | undefined
  let onPressIn = () => {}
  let onPressOut = () => {}

  if (Platform.OS === "ios") {
    onPressIn = () => {
      startPressStateColorAnimation(1)
    }
    onPressOut = () => {
      startPressStateColorAnimation(0)
    }
    containerStyles.push({
      backgroundColor: getColorInterpolation(mainBackgroundColor, pressedColor),
    })
  } else {
    background = TouchableNativeFeedback.Ripple(pressedColor)
  }

  return (
    <Touchable
      onPressIn={onPressIn}
      onPressOut={onPressOut}
      onPress={onClick}
      background={background}
      testID={testID}
      accessibilityLabel={testID}
    >
      <Animated.View style={containerStyles}>
        {!loading && (
          <>
            <Overlay insetLeftStep={ICON_MARGIN}>
              <View style={iconStyles}>
                <SVG localSVG={icon} tint={(style.iconTint as keyof TextColors) || "brand"} />
              </View>
            </Overlay>
            <Margin marginStep={TEXT_MARGIN}>
              <Text style={textStyles} testID={testID + ".text"} accessibilityLabel={testID + ".text"}>
                {buttonText}
              </Text>
            </Margin>
          </>
        )}
        {loading && (
          <Margin marginStep={TEXT_MARGIN}>
            <ActivityIndicator />
          </Margin>
        )}
      </Animated.View>
    </Touchable>
  )
}
