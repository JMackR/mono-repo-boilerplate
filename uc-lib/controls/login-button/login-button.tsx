import React from "react"
import { StyleSheet, css } from "aphrodite/no-important"
import { SVG } from "uc-lib/controls/image"
import { Margin, Overlay } from "uc-lib/controls/layout"
import { useFontForTextType, useColor } from "uc-lib/themes"
import { covertFontToWebStyleSheet } from "uc-lib/themes/utility/font-conversions"
import { ExtendedLoginButtonProps } from "./login-button.props"
import { ActivityIndicator } from "../activity-indicator"

const TEXT_MARGIN = 2.5
const ICON_DIMENSION = 38
const ICON_MARGIN = 0.5

/*
  The default behavior is no border, a 38x38px icon with no background, light background, dark text, and granite press text.
*/
export const LoginButton = (props: ExtendedLoginButtonProps) => {
  const { buttonText, onClick, isFullWidth, loading, style, icon, showBorder } = props

  const colors = useColor().colors
  const mainBackgroundColor = style.mainBackgroundColor || colors.crystal
  const pressedColor = style.pressedColor || colors.crystalPressed
  const hoverColor = style.hoverColor || colors.crystalHover
  const STYLES = StyleSheet.create({
    fullWidth: {
      width: "100%",
    },
    container: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      borderRadius: "4px",
      borderWidth: 1,
      padding: 0,
      ":hover": {
        cursor: "pointer",
        backgroundColor: hoverColor,
      },
      ":active": {
        backgroundColor: pressedColor,
      },
    },
    mainContainer: {
      flexDirection: "row",
      backgroundColor: mainBackgroundColor,
      borderColor: style.borderColor,
      borderWidth: style.borderColor === undefined ? 0 : 1,
    },
    iconContainer: {
      width: style.iconSize || ICON_DIMENSION,
      height: style.iconSize || ICON_DIMENSION,
      backgroundColor: style.iconBackgroundColor,
    },
    hideBorder: {
      borderWidth: 0,
    },
    icon: {
      verticalAlign: "middle",
    },
    text: {
      color: style.textColor || colors.alwaysDark,
      verticalAlign: "middle",
      display: "inline-block",
    },
  })

  const font = useFontForTextType("primaryBody1")
  const fontStyles = covertFontToWebStyleSheet(font)
  const textStyles: object[] = [fontStyles.text, STYLES.text]
  const containerStyles: object[] = [STYLES.mainContainer, STYLES.container]
  const iconStyles: object[] = [STYLES.iconContainer, STYLES.container]
  if (isFullWidth) {
    containerStyles.push(STYLES.fullWidth)
  }
  if (!showBorder) {
    containerStyles.push(STYLES.hideBorder)
  }

  return (
    <button className={css(containerStyles)} onClick={onClick}>
      {(!loading && (
        <>
          <Overlay insetLeftStep={ICON_MARGIN}>
            <div className={css(iconStyles)}>
              <SVG localSVG={icon} tint={style.iconTint || "brand"} />
            </div>
          </Overlay>
          <Margin marginStep={TEXT_MARGIN}>
            <div className={css(textStyles)}>{buttonText}</div>
          </Margin>
        </>
      )) || <ActivityIndicator />}
    </button>
  )
}
