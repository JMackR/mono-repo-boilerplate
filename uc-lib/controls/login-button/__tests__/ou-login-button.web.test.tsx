import React from "React"
import { shallow, ShallowWrapper } from "enzyme"
import renderer from "react-test-renderer"
import { SocialAppleIcon } from "uc-lib"
import { LoginButton } from "../login-button"

const BUTTON_TEXT = "This is a login button"
const STYLE = {
  mainBackgroundColor: "#010101",
  borderColor: "#121212",
  pressedColor: "#343434",
  textColor: "#454545",
}
const ICON = SocialAppleIcon

describe("<LoginButton />", () => {
  const createComponent = (props) => <LoginButton {...props} />
  const createWrapper = (props) => shallow(<LoginButton {...props} />)
  let wrapper: ShallowWrapper
  it('should contain "This is a login button" text', () => {
    wrapper = createWrapper({
      buttonText: BUTTON_TEXT,
      style: STYLE,
      icon: ICON,
      onClick: () => {},
    })
    expect(wrapper.html()).toEqual(expect.stringContaining(BUTTON_TEXT))
  })
  describe("Snapshot", () => {
    it("should render properly with default width", () => {
      const tree = renderer.create(
        createComponent({
          buttonText: BUTTON_TEXT,
          style: STYLE,
          icon: ICON,
          onClick: () => {},
        }),
      )
      expect(tree).toMatchSnapshot()
    })
    it("should render properly with enabling full width", () => {
      const tree = renderer.create(
        createComponent({
          buttonText: BUTTON_TEXT,
          style: STYLE,
          icon: ICON,
          onClick: () => {},
          isFullWidth: true,
        }),
      )
      expect(tree).toMatchSnapshot()
    })
  })
})
