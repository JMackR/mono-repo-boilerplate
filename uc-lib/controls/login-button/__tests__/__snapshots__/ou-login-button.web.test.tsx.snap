// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginButton /> Snapshot should render properly with default width 1`] = `
.mainContainer-container-hideBorder {
  flex-direction: row;
  background-color: #010101;
  border-color: #121212;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 4px;
  padding: 0px;
  border-width: 0px;
}

.mainContainer-container-hideBorder:hover {
  cursor: pointer;
  background-color: #F7F7F7;
}

.mainContainer-container-hideBorder:active {
  background-color: #343434;
}

.iconContainer-container {
  width: 38px;
  height: 38px;
  background-color: undefined;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 4px;
  border-width: 1px;
  padding: 0px;
}

.iconContainer-container:hover {
  cursor: pointer;
  background-color: #F7F7F7;
}

.iconContainer-container:active {
  background-color: #343434;
}

.text-text {
  font-size: 16px;
  font-weight: bold;
  font-family: Lato, sans-serif;
  color: #454545;
  vertical-align: middle;
  display: inline-block;
}

.ou {
  display: flex;
  position: absolute;
  top: undefined;
  bottom: undefined;
  left: 2px;
  right: undefined;
  flex-basis: auto;
  flex-grow: 0;
  flex-shrink: 1;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  background-color: undefined;
  z-index: 1;
  height: undefined;
  width: undefined;
}

.oumargin {
  margin: 10px;
  margin-left: undefined;
  margin-right: undefined;
  margin-top: undefined;
  margin-bottom: undefined;
  display: flex;
  position: relative;
  flex-basis: auto;
  flex-grow: 0;
  flex-shrink: 1;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  background-color: undefined;
  height: undefined;
  max-width: undefined;
  width: undefined;
}

<button
  className="mainContainer-container-hideBorder"
  onClick={[Function]}
>
  <div
    className="ou"
  >
    <div
      className="iconContainer-container"
    >
      <apple.svg
        alwaysdark="#121212"
        alwayslight="#FFFFFF"
        basalt="#8A8A8A"
        crystal="#FFFFFF"
        crystalhover="#F7F7F7"
        crystalpressed="#EEEEEE"
        disabled="#FAFAFA"
        glacialblue="#3D98D2"
        granite="#ABABAB"
        larchyellow="#F4BE19"
        larchyellowhover="#F5D624"
        larchyellowpressed="#F3A70C"
        limestone="#E2E2E2"
        limestonehover="#EEEEEE"
        limestonepressed="#D0D0D0"
        obsidian="#121212"
        overlay="rgba(18,18,18,0.3)"
        paintbrushred="#E05666"
        palousegreen="#00A87E"
        palousegreenhighlight="#DEF3EC"
        palousegreenhover="#43B893"
        palousegreenpressed="#00986A"
        quartz="#FAFAFA"
        tintcolor="#00A87E"
      />
    </div>
  </div>
  <div
    className="oumargin"
  >
    <div
      className="text-text"
    >
      This is a login button
    </div>
  </div>
</button>
`;

exports[`<LoginButton /> Snapshot should render properly with enabling full width 1`] = `
.iconContainer-container {
  width: 38px;
  height: 38px;
  background-color: undefined;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 4px;
  border-width: 1px;
  padding: 0px;
}

.iconContainer-container:hover {
  cursor: pointer;
  background-color: #F7F7F7;
}

.iconContainer-container:active {
  background-color: #343434;
}

.text-text {
  font-size: 16px;
  font-weight: bold;
  font-family: Lato, sans-serif;
  color: #454545;
  vertical-align: middle;
  display: inline-block;
}

.ou {
  display: flex;
  position: absolute;
  top: undefined;
  bottom: undefined;
  left: 2px;
  right: undefined;
  flex-basis: auto;
  flex-grow: 0;
  flex-shrink: 1;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  background-color: undefined;
  z-index: 1;
  height: undefined;
  width: undefined;
}

.oumargin {
  margin: 10px;
  margin-left: undefined;
  margin-right: undefined;
  margin-top: undefined;
  margin-bottom: undefined;
  display: flex;
  position: relative;
  flex-basis: auto;
  flex-grow: 0;
  flex-shrink: 1;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  background-color: undefined;
  height: undefined;
  max-width: undefined;
  width: undefined;
}

.mainContainer-container-fullWidth-hideBorder {
  flex-direction: row;
  background-color: #010101;
  border-color: #121212;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 4px;
  padding: 0px;
  width: 100%;
  border-width: 0px;
}

.mainContainer-container-fullWidth-hideBorder:hover {
  cursor: pointer;
  background-color: #F7F7F7;
}

.mainContainer-container-fullWidth-hideBorder:active {
  background-color: #343434;
}

<button
  className="mainContainer-container-fullWidth-hideBorder"
  onClick={[Function]}
>
  <div
    className="ou"
  >
    <div
      className="iconContainer-container"
    >
      <apple.svg
        alwaysdark="#121212"
        alwayslight="#FFFFFF"
        basalt="#8A8A8A"
        crystal="#FFFFFF"
        crystalhover="#F7F7F7"
        crystalpressed="#EEEEEE"
        disabled="#FAFAFA"
        glacialblue="#3D98D2"
        granite="#ABABAB"
        larchyellow="#F4BE19"
        larchyellowhover="#F5D624"
        larchyellowpressed="#F3A70C"
        limestone="#E2E2E2"
        limestonehover="#EEEEEE"
        limestonepressed="#D0D0D0"
        obsidian="#121212"
        overlay="rgba(18,18,18,0.3)"
        paintbrushred="#E05666"
        palousegreen="#00A87E"
        palousegreenhighlight="#DEF3EC"
        palousegreenhover="#43B893"
        palousegreenpressed="#00986A"
        quartz="#FAFAFA"
        tintcolor="#00A87E"
      />
    </div>
  </div>
  <div
    className="oumargin"
  >
    <div
      className="text-text"
    >
      This is a login button
    </div>
  </div>
</button>
`;
